### ISSUE TITLE
Premium modules

### ISSUE THREAD

**Issue by sholladay** [sholladay/fundraising-for-hapi#2]
This idea is to have a collection of modules which are pay-to-use. Generally, a premium module would be defined as such based on it being needed for production readiness and the scale of applications that need it being relatively large.

For example, API rate-limiting is a common need for public, production applications that are also popular or at-risk for abuse. However, many apps do not need API rate-limiting, such as those that are for internal use only, or those that are still in development, or some that use hapi for rendering websites without an API, etc. This makes rate-limiting a good candidate for a premium module, as its use case is fairly specific and anyone who needs it is likely making money in the course of using it.

Some other candidates:
 - Server-side caching (catbox) - only needed to improve performance once responses become computationally expensive and client-side caching is not enough, which typically implies a large-scale usage
 - Telementry / metrics - mostly used for monitoring production operations
 - OAuth - used for login/signup via external systems, which is only applicable to some apps, and is often used to improve security or for the convenience of not managing an in-house database
 - Role-based access control - used for relatively complex apps that have multiple types of users, which often implies a source of revenue or business use
 - Integrations with 3rd party products and SaaS tools (e.g. Datadog, HashiCorp Vault, etc.) - mostly used to enhance production monitoring, deployment, and security

_Thanks to @aalimovs for mentioning this in Slack._

**Comment by aalimovs** [comment]
Thanks @sholladay. There's ongoing discussion at https://hapihour.slack.com/archives/C015JQPME23/p1593087162209300.

TLDR is that I strongly believe hapi has everything it needs to become commercially successful business and earn Eran bags of money that he rightfully deserves for everything he's done to hapi and the Node.js ecosystem.

I keep mentioning https://sidekiq.org/ as an example, because I believe it's close to my understanding of Eran's desire to keep the operation small. Sidekiq was making $80k/mo in 2016 as a single person business, and still [making $160k/mo ($2 million anually) in 2020 as a one-man band](https://news.ycombinator.com/item?id=21923493).

**Comment by damusix** [comment]
Love this! Here are some ideas I can think of off the top of my head because I am planning to use these for my clients.

Ongoing services:
- Logstash / Telegraf plugins - Hook up your hapi app with your telemetry systems
- ABAC - Sometimes you want some granular permissions beyond RBAC

One-offs:
- Auth0 / Okta / Cognito integration - I'd pay a premium to save myself 2 - 4 weeks of time doing a proper integration
- Kong integration - This falls under the previous mentioned rate limiting service
- Stripe server-side integration - For subscription services, product sales, etc; with functionality to protect routes based on subscription status

**Comment by damusix** [comment]
How would this be financially modeled? There needs to be an incentive for the people developing these plugins / services. 50% to the continuing of the services + 50% directly to the development of Hapi?

**Comment by sholladay** [comment]
I think there should be two options:
 - A one time payment for individual modules (maybe $40-$50)
 - A subscription that includes all modules, including future modules (maybe $200-$300 per year)
